dali/operators/audio/resample.cc

```diff
@@ -0,0 +1,168 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <vector>
+#include "dali/operators/audio/resample.h"
+#include "dali/operators/audio/resampling_params.h"
+#include "dali/kernels/kernel_params.h"
+
+namespace dali {
+
+DALI_SCHEMA(experimental__AudioResample)
+  .DocStr(R"(Resamples a signal with a different sampling rate.
+
+The resampling is achieved by applying a sinc filter with Hann window with an extent
+controlled by the ``quality`` argument.
+
+The resampling ratio can specified either directly or as a ratio of target to source sampling rate
+or calculated from the ratio of requested output length to input length.
+)")
+  .NumInput(1)
+  .NumOutput(1)
+  .AddOptionalArg<float>("in_rate", R"(Input sampling rate.
+
+The sampling rate of the input sample. This parameter must be specified together with ``out_rate``.
+The value is relative to ``out_rate`` and doesn't need to use any specific unit as long as the
+units of input and output rates match.
+
+The ``in_rate`` and ``out_rate`` parameters cannot be specified together with ``scale`` or
+``out_length``.)",
+    nullptr, true)
+  .AddOptionalArg<float>("out_rate", R"(Input sampling rate.
+
+The sampling rate of the input sample. This parameter must be specified together with ``out_rate``.
+The value is relative to ``out_rate`` and doesn't need to use any specific unit as long as the
+units of input and output rates match.
+
+The ``in_rate`` and ``out_rate`` parameters cannot be specified together with ``scale`` or
+``out_length``.)",
+    nullptr, true)
+  .AddOptionalArg<float>("scale", R"(The scaling factor.
+
+The scaling factor is the ratio of the target sampling rate to source sampling rate. For example,
+a ``scale=2`` will produce an output with twice as many samples as the input.
+
+This parameter cannot be specified together with ``in_rate`` and ``out_rate`` or ``out_length``.)",
+    nullptr, true)
+  .AddOptionalArg<int64_t>("out_length", R"(The requested output length, in samples.
+
+The scaling factor is the ratio of this output length to the input length. This parameter
+cannot be specified together with ``in_rate``, ``out_rate`` or ``scale``.)",
+    nullptr, true)
+  .AddOptionalArg("quality", R"(Resampling quality, where 0 is the lowest, and 100 is
+the highest.
+
+0 gives 3 lobes of the sinc filter, 50 gives 16 lobes, and 100 gives 64 lobes.)",
+    50.0f, false)
+  .AddOptionalArg<DALIDataType>("dtype", R"(The ouput type.
```

Comment:
I don't know if this is clear enough here, that the input should be either -1, 1 for floats, or full range for integers?

Reply:
Well, -1..1 for integers makes _very_ little sense. I could add examples.

Reply:
Up to you. On the other hand we don't do that for other ops....